``` {r setup, echo=FALSE, message=FALSE, include=FALSE, error=FALSE}
library(GenomicRanges, warn.conflicts=F)
library(magrittr)
library(parallel)
library(ggplot2)
library(Gviz)
library(BSgenome.Dmelanogaster.UCSC.dm6)
library(TxDb.Dmelanogaster.UCSC.dm6.ensGene)

setwd("/papagianni_PNAS_2017/analysis/")
options(knitr.figure_dir = "chipnexus_single_enhancer_example")

source("shared_code/knitr_common.r")
source("shared_code/granges_common.r")
source("shared_code/ggplot_common.r")
source("shared_code/metapeak_common.r")
source("shared_code/sample_common.r")
```


# ChIP-nexus single enhancer example

**Author:** [Wanqing Shao](mailto:was@stowers.org)

**Generated:** `r format(Sys.time(), "%a %b %d %Y, %I:%M %p")`

ChIP-nexus single gene example. Both zoomed in and zoomed out version will be included. 

```{r single_gene_zoomed, fig.height=8}
enhancer <- read.csv("region_of_interests.csv", header= T, stringsAsFactors = F)
enhancer_gr <- makeGRangesFromDataFrame(enhancer, keep.extra.columns = T)

enhancer_zoomed <- enhancer_gr[enhancer_gr$type == "zoomed"]
enhancer_whole <- enhancer_gr[enhancer_gr$type == "whole"]


motif_gr_list <- list(dl_2mm = readRDS("rdata/dl_motif_GGRWWTTCC_2_mm_cl.rds")$gr,
                      cic_0mm = readRDS("rdata/cic_motif_TSAATGAA_no_mm_cl.rds")$gr,
                      cic_1mm = readRDS("rdata/cic_motif_TSAATGAA_1_mm_only_cl.rds")$gr)

samples <- list(tl_dl = load_bigwig("tl_dl_nexus"),
                tl_cic = load_bigwig("tl_cic_nexus"),
                gd_cic = load_bigwig("gd_cic_nexus"))

build_nexus_track <- function(data, name, chr, start, end, ylim, pos_col, neg_col){
  data_track <- DataTrack(data=data, 
                          start=start:end, width=0, 
                          chromosome=chr,
                          genome="dm6", 
                          name=name,
                          type="histogram",
                          groups=c("pos", "neg"),
                          ylim=c(-ylim, ylim),
                          legend=FALSE,
                          col=c(pos_col, neg_col),  
                          col.histogram ="transparent",
                          background.title = "white")
  data_track
}

build_gene_track <- function(chr, start, end){
    gtrack <- GeneRegionTrack(range=TxDb.Dmelanogaster.UCSC.dm6.ensGene,
                             genome="dm6",
                             chromosome=chr,
                             name="Genes",
                             showId=TRUE,
                             geneSymbol=TRUE,
                             fill="gray60",
                             background.title = "white",
                             start=start, end=end,
                             collapseTranscripts=FALSE) 
    gtrack
}

build_motif_track <- function(motif_gr,chr, start, end, name, col){
  motif_gr <- motif_gr[seqnames(motif_gr) == chr & 
                       start(motif_gr) >= start - 10 &
                       end(motif_gr) <= end + 10  ]
  motif_track <- GeneRegionTrack(range=motif_gr,
                                 genome="dm6",
                                 chromosome=chr,
                                 start = start, end = end,
                                 name=name,
                                 fill=col,
                                 col=col, col.grid=col,
                                 background.title = "white",
                                 collapseTranscripts=FALSE) 
}


cic_nexus_col <- rev(c("#c6331d", "#D05947"))
dl_nexus_col <- rev(c("#27679C", "#3C76A5"))

motif_colors <- c(dl_1mm = "#276599",
                  cic_0mm = "#D02309",
                  cic_1mm = "#F98978")


build_track_plot<- function(region.gr, nexus_sample_list, motif_list = NULL) {
  
  r_chr <- as.character(seqnames(region.gr))
  r_start <- start(region.gr)
  r_end   <- end(region.gr)
  gtrack <- build_gene_track(r_chr, r_start, r_end)
  
  nexus_data <- mclapply(nexus_sample_list, function(x){
          exo_matrix <- exo_metapeak_matrix(resize(region.gr, 1, "start"), x, upstream=0, downstream = width(region.gr))
          exo_matrix_t <- rbind(exo_matrix$pos,-1 * exo_matrix$neg)
          exo_matrix_t}, mc.cores =1)
  
  dl_ylim <- max(abs(nexus_data$tl_dl))
  cic_ylim <- c(nexus_data$tl_cic, nexus_data$tl_cic) %>% abs %>% max
  
  nexus_track_list <- mapply(build_nexus_track, 
                             nexus_data, c("dl Toll10b", "cic Toll10b", "cic gd7"),
                             list(r_chr), list(r_start), list(r_end),
                             list(dl_ylim,cic_ylim,cic_ylim ),
                             c(dl_nexus_col[1], cic_nexus_col[1], cic_nexus_col[1]),
                             c(dl_nexus_col[2], cic_nexus_col[2], cic_nexus_col[2]))
  
  genometrack <- GenomeAxisTrack(range =region.gr )
  tlist <- c(list(genometrack), nexus_track_list, list(gtrack))    
  
  
  if(!is.null(motif_list)){
      motif_track_list <- mapply(build_motif_track,
                             motif_list,
                             list(r_chr), list(r_start),list(r_end),
                             c("dl motif", "cic motif no mm", "cic motif 1 mm"),
                             motif_colors) 
      
      tlist <- c(list(genometrack), nexus_track_list, motif_track_list, list(gtrack))    

  }
  tlist
}


plot_tracks <- function(tlist, region.gr, name, zoomed =F){
  r_chr <- as.character(seqnames(region.gr))
  r_start <- start(region.gr)
  r_end   <- end(region.gr)
  
  if(zoomed){
    tsizes <- c(0.2,rep(0.4, 3), rep(0.1, 3), 0.2)
    x <- plotTracks(tlist, chromosome=r_chr, from=r_start,to=r_end, sizes = tsizes, 
             cex.title=0.8, cex.axis=0.8, col.title="black", col.axis="black",
             fontcolor.legend="black", cex.legend=0.8, main = name, rotation.title = 0) 
  }else{
    tsizes <- c(0.2, rep(0.4, 3), 0.2)
    x <- plotTracks(tlist, chromosome=r_chr, from=r_start,to=r_end, sizes = tsizes, 
             cex.title=0.8, cex.axis=0.8, col.title="black", col.axis="black",
             fontcolor.legend="black", cex.legend=0.8, main = name, rotation.title = 0) 
    
  }
}


zoomed_view_tlist <- cache("zoomed_view", function(){
  lapply(enhancer_zoomed, function(x)build_track_plot(x, samples, motif_gr_list))
})
names(zoomed_view_tlist) <- enhancer_zoomed$region_name
nothing <- lapply(names(zoomed_view_tlist), function(x)plot_tracks(zoomed_view_tlist[[x]], enhancer_zoomed[enhancer_zoomed$region_name == x], paste(x, "zoomed"), T))
```

```{r single_gene_whole, fig.height=6}
whole_view_tlist <- cache("whole_view", function(){
  lapply(enhancer_whole, function(x)build_track_plot(x, samples))
})
names(whole_view_tlist) <- enhancer_whole$region_name

nothing <- lapply(names(whole_view_tlist), function(x)plot_tracks(whole_view_tlist[[x]], enhancer_whole[enhancer_whole$region_name == x], x))
```

```{r echo= F}
sessionInfo()
```